specification/2.0/appendix/performance-metrics/README.md

```diff
@@ -0,0 +1,107 @@
+# Runtime Performance Metrics 
+
+The goal of these metrics is to assess the runtime performance needs of glTF assets.  
+By runtime we mean the cpu/gpu, bandwidth and memory resources needed to render the model in realtime.  
+These performance metrics are divided in two major categories - `complexity` and `memory`
+
+## Performance Metric Categories
+
+### Complexity  
+
+The `complexity`category provides metrics to calculate cpu/gpu and memory bandwidth target requirements - larger numbers will mean a more powerful target device is needed.  
+These metrics do not provide a mapping to target device performance, this mapping must be done by users of the metrics.  
+
+### Screen space coverage
+
+The metrics do not include screen space coverage, it is advised that clients shall know how large the render area will be before requesting a model (from the backend)  
+It is up to the backend to take screen coverage into account when providing a suitable model to the client, for instance by taking width + height as parameters.  
+
+### Memory  
+
+The `memory` category provides metrics to calculate the runtime memory (size) needs of the model.  
+Using these metrics an estimate of target memory can be calculated - however keep in mind that the exact memory requirements will vary with viewer implementations.  
+
+
+
+|                 |             |          |                                                                |             |
+|-----------------|-------------|-------------|----------|----------------------------------------------------------------|
+| Name            | Type        | Category    | Property |Description                                                     |
+|[vertexCount]    | Integer     | Complexity  |Scene     |Total number of vertices used by the Nodes in a scene           |  
+|[nodeCount]      | Integer     | Complexity  |Scene     |Max nodecount in scene (add upp all nodes in a scene)           |  
+|[primitiveCount] | Integer     | Complexity  |Scene     |Total number of referenced primitives (per scene).  This figure is the un-batched number of primitives, engines may optimize if primitives and meshes share textures. |  
+|[textures]       | Integer     | Complexity  |Scene     |Flags specifying presence of materials and it's texture usage, this is the aggregated max usage. BASECOLOR, METALLICROUGHNESS, NORMAL, OCCLUSION, EMISSIVE, (SPECULARGLOSS) |  
+|[maxNodeDepth]   | Integer     | Memory      | Scene    | The max node depth of the scene, ie the max number of parent/child relations. This number will represent the max stack depth needed when traversing the nodegraph |
+|[attributes]     | Integer     | Memory      | Asset    | The total memory footprint, in bytes, for the models attribute buffer storage |  
+|[textureSizes]  |Dimension    | Memory      | Asset    | The width and height of defined textures in the Asset. |  
```

Comment:
Some suggestions on this list:

- **vertexCount:** LGTM.
- **vertexMemory:** Suggested as an alternative name to `attributes`. 
- **nodeCount:** LGTM.
- **primitiveCount:** LGTM.
- **materialCount:** Suggested addition. This value, combined with primitiveCount, heavily affects draw count and thus CPU time.
- **textures:** I'm not sure I understand this metric, could you explain further? Or see comment below.
- **textureCount:** Suggested addition. 
- **textureMemory:** Suggested alternative to `textureSizes`, which is more future-proof for compressed textures which have lower per-pixel cost.

I haven't yet seen an asset with node depth >5-10, so I don't think max stack depth is a major concern here. The `nodeCount` metric alone should be fine.

One factor that is mostly missing from this list is per-pixel cost of the materials. Notably, materials using `KHR_materials_unlit` are _much_ cheaper when filling the screen than core PBR materials, and PBR materials using more of the future PBR Next extensions will be more expensive yet. I'm not sure how to encode that in a metric, at this point. The `textures` property above sort of helps here, although I think it's less important to know what material properties are using textures, than to know what material properties are in effect at all. For example, subsurface scattering will be expensive on some clients regardless of whether a texture is used for the effect.

Reply:
^correction, I think *both* `textureMemory` and `textureMaxSize` would be useful.

Reply:
node depths can vary based on the asset types.  Eg. For character models (dragons etc), it is not uncommon to have (20+) deeper tree, to drive a range of skeleton behaviors.   In our case in Magic leap, simple (but animatable) avatars have crossed way beyond 5 level depth.

I do agree that, compared to vertex, textures etc, this node-depth conveys the metric at a relatively different scale (roughly about logarithmic to node count, in typical cases)   
Thats why maxNodeDepth was meant (at least for me) to be on the 'complexity' category.

Reply:
Thanks for your comments @donmccurdy 
I will update `attributes` to vertexMemory.

The `textures` metric is used to calculate the per pixel cost.
This will store an array of values indicating the texture/material properties of the most _complex_ material.
For a normal type of PBR material this would contain something like [BASECOLOR, METALLICROUGHNESS, NORMAL, OCCLUSION]
In future with the new extensions coming we would add something like [CLEARCOAT, DIELECTRIC_F0] - or whatever they shall be called :-)

I guess what's missing is a flag to indicate presense of PBR material and how to handle `KHR_materials_unlit` in a generic way?

Reply:
Texture count can be found by looking at the size of the array of textureSizes.
I see your point regarding `textureMemory` - I will update and add a metric for `maxTextureSize`

Reply:
Ok, after reviewing my changes I have changed back to similar to what it was first:

[textures] - the worst case texture maps/channels - this can be used to estimate worst case per pixel texel cost (bandwidth)

[materials] - the worst case computational cost per output pixel - this can be used to estimate per pixel computaton cost.
Handles pbr, speculargloss and unlit.

[accessors] - the asset vertex accessor usage, with count, componentType and type - this can be used to estimate the runtime memory consumption for vertex buffers.

[textureSizes] - the asset texture sizes and component formats - this can be used to estimate runtime texture memory.

For accessors and texture sizes I would rather store the details (count or size + format) because clients can take platform details into account.
One target may convert SHORT to FLOAT, or the other way around - a target may need to handle texture sources in a special way, eg 3 component textures are almost always stored (in GPU memory) as 4 bytes.
This way is also flexible when it comes to future use of BasisU - the texture will be transcoded to a supported target format. 
Memory consumption for target format can be calculated by width * height * target format bitdepth.

Reply:
> The `textures` metric is used to calculate the per pixel cost.

Knowing the purposes of the textures in the material would be an incomplete measure of per-pixel cost, I think. More advanced shading features (clearcoat, sheen, thin transparency) add cost simply by being enabled — whether a texture is used to achieve them or not. By comparison, the cost of a texture lookup is mostly fixed regardless of what that texture is used for.

A list of extensions might be the simplest way to express that, since the advanced PBR features will mostly be enabled by per-feature extensions.